default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1601 @@
+        #
+        self.hulls_for_planets.clear()
+        self.parts_for_planets.clear()
+        get_shipdesign = fo.getShipDesign
+        get_hulltype = fo.getHullType
+        empire = fo.getEmpire()
+        empire_id = empire.empireID
+        universe = fo.getUniverse()
+        planets_with_shipyards = _get_planets_with_shipyard()
+        available_hulls = list(empire.availableShipHulls)   # copy so we can sort it locally
+        try:
+            idx = available_hulls.index(TESTDESIGN_PREFERED_HULL)
+            available_hulls[0], available_hulls[idx] = available_hulls[idx], available_hulls[0]
+        except ValueError:
+            print "WARNING: Tried to use '%s' as testhull but it is not in available_hulls." % TESTDESIGN_PREFERED_HULL,
+            print "Please update ShipDesignAI.py according to the new content."
+            traceback.print_exc()
+        testdesign_ids = [shipDesignID for shipDesignID in empire.allShipDesigns
+                          if get_shipdesign(shipDesignID).name(False).startswith(TESTDESIGN_NAME_BASE)]
+
+        testdesign_names_hull = [get_shipdesign(ID).name(False) for ID in testdesign_ids
+                                 if get_shipdesign(ID).name(False).startswith(TESTDESIGN_NAME_HULL)]
+        testdesign_names_part = [get_shipdesign(ID).name(False) for ID in testdesign_ids
+                                 if get_shipdesign(ID).name(False).startswith(TESTDESIGN_NAME_PART)]
+
+        available_slot_types = {slottype for slotlist in [get_hulltype(hull).slots for hull in available_hulls]
+                                for slottype in slotlist}
+
+        new_parts = [_get_part_type(part) for part in empire.availableShipParts
+                     if part not in self.strictly_worse_parts]
+        pid = self.production_cost.keys()[0]  # TODO: Check if actually location invariant
+        for new_part in new_parts:
+            self.strictly_worse_parts[new_part.name] = []
+            for part_class in ALL_META_CLASSES:
+                if new_part.partClass in part_class:
+                    for old_part in [_get_part_type(part) for part in self.strictly_worse_parts
+                                     if part != new_part.name]:
+                        if old_part.partClass in part_class:
+                            if new_part.capacity >= old_part.capacity:
+                                a = new_part
+                                b = old_part
+                            else:
+                                a = old_part
+                                b = new_part
+                            if (self.production_cost[pid][a.name] <= self.production_cost[pid][b.name]
+                                    and {x for x in a.mountableSlotTypes} >= {x for x in b.mountableSlotTypes}):
+                                # ToDo: Maybe add production_time as additional condition?
+                                self.strictly_worse_parts[a.name].append(b.name)
```

Comment:
it looks like this manner of storing "strictly_worse" info inherently assumes that all the cost, etc.,  is location invariant.  Of course, for now that is not a terrible assumption, and I don't think that the invariance check is even exposed to python yet, but I think it would be better to frame things as this whole strictly-worse tracking will only be done for parts that are cost-time location-invariant.  So just some comments to that effect are probably enough for now.